### PR TITLE
Homography example

### PR DESCRIPTION
**Pull request by fantaosha**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: examples/homography_estimation.py [hunk 1]
```diff
@@ -0,0 +1,463 @@
+import kornia
```
Comment (luisenp) [inline examples/homography_estimation.py]: Need to add copyright notice here.

File: examples/homography_estimation.py [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+            img2 = torch.clamp(img2, 0.0, 1.0)
+            img1 = self.rpa.forward(img1)
+            img2 = self.rpa.forward(img2)
+
+        data = {"img1": img1[0], "img2": img2[0], "H_1_2": H_1_2[0]}
+
+        return data
+
+
+def grid_sample(image, optical):
```
Comment (luisenp) [inline examples/homography_estimation.py:125]: If this is code we got from somewhere else, we should add a link to the source.  And probably move it to a third party folder, since we shouldn't be claiming copyright on this. @ddetone @joeaortiz 

Does this also apply to any of the other code below?

File: examples/homography_estimation.py [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+
+
+def warp_perspective_norm(H, img):
+    height, width = img.shape[-2:]
+    grid = kornia.utils.create_meshgrid(
+        height, width, normalized_coordinates=True, device=H.device
+    )
+    Hinv = torch.inverse(H)
+    warped_grid = kornia.geometry.transform.homography_warper.warp_grid(grid, Hinv)
+    # grid_sample = torch.nn.functional.grid_sample
```
Comment (luisenp) [inline examples/homography_estimation.py:187]: Remove commented out code.

File: examples/homography_estimation.py [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+            os.system(cmd)
+            os.makedirs(dataset_path)
+            cmd = "tar -xf %s/%s -C %s" % (dataset_root, tar, dataset_path)
+            print("Running command: ", cmd)
+            os.system(cmd)
+
+    return dataset_paths
+
+
+class HomographyDataset(Dataset):
```
Comment (luisenp) [inline examples/homography_estimation.py:57]: Can we add some brief comments to all functions describing what they do? (both in/out of this class). Nothing too elaborate, just a couple of sentences. @ddetone @joeaortiz

File: examples/homography_estimation.py [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+    viz_every = 10
+    disp_every = 10
+    save_every = 100
+    num_epochs = 999
+    outer_lr = 1e-4
+    use_cnn = True
+
+    log_dir = os.path.join(os.getcwd(), "viz")
+    os.makedirs(log_dir, exist_ok=True)
+    print("Writing to tensorboard at", os.getcwd())
```
Comment (luisenp) [inline examples/homography_estimation.py:335]: I don't think we need the tensorboard dependency, we can just log the loss to a csv file or something similar.

File: examples/homography_estimation.py [hunk 6]
```diff
@@ -0,0 +1,463 @@
...
+
+        for _, data in enumerate(dataloader):
+
+            outer_optim.zero_grad()
+
+            img1 = data["img1"].to(device)
+            img2 = data["img2"].to(device)
+            Hgt_1_2 = data["H_1_2"].to(device)
+
+            objective = th.Objective()
```
Comment (luisenp) [inline examples/homography_estimation.py:365]: Is there a reason why we create the objective and theseus layer inside the loop? Couldn't we create the TheseusLayer outside, around the part where we create the cnn_model, and just populate the data using the forward method?

**Review by luisenp — CHANGES_REQUESTED** [review]
Thanks everyone for the hard work on this!! Left some comments, mostly organization stuff.

File: examples/homography_estimation.py (hunk 4 above)
Comment (mhmukadam) [reply examples/homography_estimation.py:57]: Remember to also appropriately update the examples/readme file.

File: examples/homography_estimation.py (hunk 1 above)
Comment (fantaosha) [reply examples/homography_estimation.py]: Added.

File: examples/homography_estimation.py (hunk 3 above)
Comment (fantaosha) [reply examples/homography_estimation.py:187]: Removed.

File: examples/homography_estimation.py (hunk 5 above)
Comment (fantaosha) [reply examples/homography_estimation.py:335]: @ddetone @joeaortiz

File: examples/homography_estimation.py (hunk 6 above)
Comment (fantaosha) [reply examples/homography_estimation.py:365]: @ddetone @joeaortiz

File: examples/homography_estimation.py (hunk 5 above)
Comment (joeaortiz) [reply examples/homography_estimation.py:338]: Sounds good to me, replacing tensorboard with writing to a csv

File: examples/homography_estimation.py (hunk 6 above)
Comment (joeaortiz) [reply examples/homography_estimation.py:368]: Yes, I agree with Luise, we should create the Theseus layer outside of the loop. I think it was put there before because the loop was over different problem formulations but should definitely be moved outside now.

File: examples/homography_estimation.py (hunk 4 above)
Comment (joeaortiz) [reply examples/homography_estimation.py:63]: I've added some comments now describing the functions.

File: examples/homography_estimation.py (hunk 5 above)
Comment (joeaortiz) [reply examples/homography_estimation.py]: I removed the logging of the loss altogether as it is being printed anyway.

**Review by mhmukadam — APPROVED** [review]
Looks great! Thank you @joeaortiz, @ddetone, @fantaosha and @luisenp for working on wrapping up this example.

**Review by mhmukadam — APPROVED** [review]
